### PR TITLE
fix/pfsnmp_queue_distribute

### PR DESCRIPTION
**Pull request by jrouzierinverse**
Description
===========
Distribute pfsnmp trap jobs between queues based off switch id.
Where each job will go to the same queue if it is for the same switch id.

Impacts
=======
pfqueue and snmp

Delete branch after merge
=========================
NO

Checklist
=========
- [ ] Document the feature
- [x] Add unit tests
- [ ] Add acceptance tests (TestLink)

### CODE REVIEW

**Comment by fdurand** [comment]
Just appied thus PR on a setup.
Just wait to see if it works, if yes i will merge it.

**Comment by fdurand** [comment]
Doesn't work, here the error:

```
packetfence: FATAL snmp_perl_trapd.pl(2241): Undefined subroutine &pf::pfqueue::producer::redis::hash_queue_name called at /usr/local/pf/lib/pf/pfqueue/producer/redis.pm line 134
```

**Comment by jrouzierinverse** [comment]
Fixed

**Comment by fdurand** [comment]
Now:

```
FATAL snmp_perl_trapd.pl(7937): Illegal modulus zero at /usr/local/pf/lib/pf/pfqueue/producer/redis.pm line 102.
```

File: lib/pf/task/pfsnmp.pm [hunk 1]
```diff
@@ -892,7 +893,7 @@ requeueTrap
 sub requeueTrap {
     my ($self, $args) = @_;
     my $client = pf::pfqueue::producer::redis->new();
-    $client->submit("pfsnmp", "pfsnmp", $args);
+    $client->submit_hashed($ConfigPfqueue{pfsnmp}{workers}, $args->{switchId}, "pfsnmp", "pfsnmp", $args);
```
Comment (fdurand) [inline lib/pf/task/pfsnmp.pm]: $ConfigPfqueue{queue_config}{pfsnmp}{workers}

File: lib/pf/task/pfsnmp_parsing.pm [hunk 2]
```diff
@@ -92,7 +93,7 @@ sub doTask {
     }
 
     my $client = pf::pfqueue::producer::redis->new(queue => 'pfsnmp');
-    $client->submit("pfsnmp", "pfsnmp", $trap);
+    $client->submit_hashed($ConfigPfqueue{pfsnmp}{workers}, $switch_id, "pfsnmp", "pfsnmp", $trap);
```
Comment (fdurand) [inline lib/pf/task/pfsnmp_parsing.pm]: $ConfigPfqueue{queue_config}{pfsnmp}{workers}

**Comment by jrouzierinverse** [comment]
Fixed

**Comment by fdurand** [comment]
Still getting:

```
packetfence: FATAL snmp_perl_trapd.pl(15455): Illegal modulus zero at /usr/local/pf/lib/pf/pfqueue/producer/redis.pm line 102.
```

File: lib/pf/snmptrapd.pm [hunk 3]
```diff
@@ -42,7 +44,7 @@ sub receiver {
         redis => _redis_client(),
     });
 #    Delay parsing by two seconds to allow snmp to do it's magic
-    $producer->submit_delayed("pfsnmp_parsing", "pfsnmp_parsing", 2000, [$trapInfo, $variables]);
+    $producer->submit_delayed_hashed($WORKER, $switchIp, "pfsnmp_parsing", "pfsnmp_parsing", 2000, [$trapInfo, $variables]);
```
Comment (fdurand) [inline lib/pf/snmptrapd.pm:47]: $WORKER is not set.

Comment (jrouzierinverse) [reply lib/pf/snmptrapd.pm:47]: Did you copy the var/conf/snmptrapd.conf?

**Comment by fdurand** [comment]
Works now, thanks.

**Pull request merged**